File: src/main/java/com/zunza/buythedip/cryptocurrency/client/CoinMarketCapClient.java
package com.zunza.buythedip.cryptocurrency.client;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

import com.zunza.buythedip.cryptocurrency.dto.cmc.CryptoCurrencyMetadataDto;

@Component
public class CoinMarketCapClient {

	private final WebClient webClient;

	private static final String URL = "https://pro-api.coinmarketcap.com";

	public CoinMarketCapClient(
		@Value("${coinMarketCap.key}")
		String apiKey,
		WebClient.Builder webClientBuilder
		) {
		this.webClient = webClientBuilder
			.baseUrl(URL)
			.defaultHeader("X-CMC_PRO_API_KEY", apiKey)
			.build();
	}

	public CryptoCurrencyMetadataDto getCryptoCurrencyMetadata(String symbolParam) {
		return webClient.get()
			.uri(uriBuilder -> uriBuilder
				.path("/v2/cryptocurrency/info")
				.queryParam("skip_invalid", "true")
				.queryParam("symbol", symbolParam)
				.build()
			)
			.retrieve()
			.bodyToMono(CryptoCurrencyMetadataDto.class)
			.block();
	}
}

File: src/main/kotlin/com/zunza/buythedip/external/coinmarketcap/client/CoinMarketCapClient.kt
package com.zunza.buythedip.external.coinmarketcap.client

import com.zunza.buythedip.external.coinmarketcap.dto.Metadata
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import kotlin.jvm.java

@Component
class CoinMarketCapClient(
    @Value("\${coinMarketCap.key}") private val apiKey: String,
    webClientBuilder: WebClient.Builder
) {
    companion object {
        private const val BASE_URL = "https://pro-api.coinmarketcap.com"
    }

    private val webClient = webClientBuilder
        .baseUrl(BASE_URL)
        .defaultHeader("X-CMC_PRO_API_KEY", apiKey)
        .codecs { config -> config
            .defaultCodecs()
            .maxInMemorySize(19 * 1024 * 1024)
        }
        .build();

    fun getInfo(symbols: String): Metadata {
        return webClient.get()
            .uri { uriBuilder -> uriBuilder
                .path("/v2/cryptocurrency/info")
                .queryParam("skip_invalid", "true")
                .queryParam("symbol", symbols)
                .build()
            }
            .retrieve()
            .bodyToMono(Metadata::class.java)
            .block()!!
    }
}
